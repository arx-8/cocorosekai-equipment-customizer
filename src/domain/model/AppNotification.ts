import { OptionsObject, SnackbarMessage } from "notistack"

export type AppNotification = {
  dismissed?: boolean
  key: Required<OptionsObject>["key"]
  message: SnackbarMessage
  options?: OptionsObject
}
