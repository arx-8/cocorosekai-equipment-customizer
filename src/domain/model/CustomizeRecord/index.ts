import {
  Attribute,
  EquipCost,
  Equipment,
  EquipmentId,
  HitPointNum,
  MagicAttackNum,
  MagicDefenseNum,
  PhysicalAttackNum,
  PhysicalDefenseNum,
  SpecialEffect,
} from "src/domain/model/Equipment"
import { toUniq } from "src/utils/arrayUtils"

/**
 * 画面上の選択値は「未選択」の位置もあるため、 undefined able
 */
export type SelectedEquipmentIds = (EquipmentId | undefined)[]

export type CustomizeRecord = {
  customizeMemo: string
  equippedIds: SelectedEquipmentIds
  mixedAttributes: Attribute[]
  mixedSpecialEffects: SpecialEffect[]
  totalEquipCost: EquipCost
  totalSpecValue: number
  totalStatuses: {
    hp: HitPointNum
    magicAtk: MagicAttackNum
    magicDef: MagicDefenseNum
    physicalAtk: PhysicalAttackNum
    physicalDef: PhysicalDefenseNum
  }
}

/**
 * 1 編成あたりに装備できる数は決まっているため
 */
export type CellColIndex = 0 | 1 | 2 | 3 | 4 | 5
export const MAX_EQUIPMENTS_NUM = 6

export type CellIndex = {
  colIndex: CellColIndex
  rowIndex: number
}

export const calcCustomizeRecord = (
  equipments: Equipment[]
): Omit<CustomizeRecord, "equippedIds" | "customizeMemo"> => {
  return equipments.reduce(
    (acc, curr) => {
      return {
        // TODO totalSpecValue の計算ロジックの実装
        mixedAttributes: toUniq([...acc.mixedAttributes, curr.attribute]),
        mixedSpecialEffects: toUniq([
          ...acc.mixedSpecialEffects.map((e) => e.rawText),
          ...curr.specialEffects.map((e) => e.rawText),
        ]).map((rawText) => {
          return { rawText }
        }),
        totalEquipCost: (acc.totalEquipCost + curr.equipCost) as EquipCost,
        totalSpecValue: 0,
        totalStatuses: {
          hp: (acc.totalStatuses.hp + curr.statuses.hp) as HitPointNum,
          magicAtk: (acc.totalStatuses.magicAtk +
            curr.statuses.magicAtk) as MagicAttackNum,
          magicDef: (acc.totalStatuses.magicDef +
            curr.statuses.magicDef) as MagicDefenseNum,
          physicalAtk: (acc.totalStatuses.physicalAtk +
            curr.statuses.physicalAtk) as PhysicalAttackNum,
          physicalDef: (acc.totalStatuses.physicalDef +
            curr.statuses.physicalDef) as PhysicalDefenseNum,
        },
      }
    },
    {
      mixedAttributes: [] as Attribute[],
      mixedSpecialEffects: [] as SpecialEffect[],
      totalEquipCost: 0 as EquipCost,
      totalSpecValue: 0,
      totalStatuses: {
        hp: 0 as HitPointNum,
        magicAtk: 0 as MagicAttackNum,
        magicDef: 0 as MagicDefenseNum,
        physicalAtk: 0 as PhysicalAttackNum,
        physicalDef: 0 as PhysicalDefenseNum,
      },
    }
  )
}
