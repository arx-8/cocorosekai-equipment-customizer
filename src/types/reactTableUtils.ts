import {
  ColumnInstance,
  TableInstance,
  UseFiltersColumnOptions,
  UseFiltersColumnProps,
  UseFiltersInstanceProps,
  UseSortByColumnOptions,
  UseSortByColumnProps,
  UseSortByInstanceProps,
  UseTableColumnOptions,
} from "react-table"

/**
 * for react-table
 */

/**
 * 使う PluginHook によって、型を合成する必要がある
 */
export type TableInstanceOverride<TData extends object> = TableInstance<TData> &
  UseFiltersInstanceProps<TData> &
  UseSortByInstanceProps<TData>

export type ColumnInstanceOverride<TData extends object> = ColumnInstance<
  TData
> &
  UseFiltersColumnProps<TData> &
  UseSortByColumnProps<TData>

export type ColumnOptionsOverride<TData extends object> = UseTableColumnOptions<
  TData
> &
  UseFiltersColumnOptions<TData> &
  UseSortByColumnOptions<TData>

export type FilterRendererProps<TData extends object = {}> = {
  column: ColumnInstance<TData> & UseFiltersColumnProps<TData>
}
