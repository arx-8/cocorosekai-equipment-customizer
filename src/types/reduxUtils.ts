import { Action, AnyAction } from "redux"
import { ThunkAction, ThunkDispatch } from "redux-thunk"
import { RootState } from "src/store/store"

/**
 * Type utility for redux
 */

/**
 * redux-thunk dispatch shorthand for this application.
 *
 * State と extraArgument は、applyMiddleware の時点で決定するため、決め打ち
 * Action は型定義するコスパが悪い(ActionCreatorを使うため、タイポや未定義の可能性は低い)ため、AnyAction
 */
export type AppThunkDispatch = ThunkDispatch<RootState, void, AnyAction>

/**
 * redux-thunk action shorthand for this application.
 *
 * ThunkAction の <R> を void | Promise<void> に制約してる理由は、実装の複雑化を避けるため。
 * 「Promise.resolve の結果を受け取って Component 側でロジックを実装」することを防ぐ。
 * 基本的に、return の Promise すら使うべきではないため、型引数省略時のデフォルト値は void としている。
 *
 * @template TReturn Type of return
 * @template TAction Type of acceptable action
 */
export type AppThunkAction<
  TReturn extends void | Promise<void> = void,
  TAction extends Action = AnyAction
> = ThunkAction<TReturn, RootState, void, TAction>
