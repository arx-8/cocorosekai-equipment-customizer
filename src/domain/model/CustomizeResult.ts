import {
  Attribute,
  EquipCost,
  EquipmentId,
  HitPointNum,
  MagicAttackNum,
  MagicDefenseNum,
  PhysicalAttackNum,
  PhysicalDefenseNum,
  SpecialEffect,
} from "src/domain/model/Equipment"
import { Brand } from "src/types/tsUtils"

export type CustomizeResult = {
  rowNo: CustomizeResultRowNo
  equippedIds: EquipmentId[]
  totalSpecValue: number
  totalEquipCost: EquipCost
  mixedAttributes: Attribute[]
  totalStatuses: {
    hp: HitPointNum
    magicAtk: MagicAttackNum
    magicDef: MagicDefenseNum
    physicalAtk: PhysicalAttackNum
    physicalDef: PhysicalDefenseNum
  }
  mixedSpecialEffects: SpecialEffect[]
}

type CustomizeResultRowNo = Brand<number, "CustomizeResultRowNo">
