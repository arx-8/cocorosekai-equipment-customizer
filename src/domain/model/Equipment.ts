import { data } from "src/assets/data"
import { Brand } from "src/types/tsUtils"

export type Equipment = {
  id: EquipmentId
  name: string
  rawName: string
  equipCost: EquipCost
  attribute: Attribute
  statuses: {
    hp: HitPointNum
    magicAtk: MagicAttackNum
    magicDef: MagicDefenseNum
    physicalAtk: PhysicalAttackNum
    physicalDef: PhysicalDefenseNum
  }
  imageUrl: string | null
  specialEffects: SpecialEffect[]
  // image:
  /** 強化回数 (+ の数) */
  enhancementsNum: 0 | 1 | 2
}

/** 装備 ID (内部ロジック用) */
export type EquipmentId = Brand<number, "EquipmentId">

/** 属性 */
export type Attribute =
  // 無
  | "void"
  // 火
  | "fire"
  // 水
  | "water"
  // 風
  | "wind"
  // 光
  | "light"
  // 闇
  | "darkness"

/** 装備のためのコスト */
export type EquipCost = Brand<number, "Cost">

/** 物理 攻撃 (数値) */
export type PhysicalAttackNum = Brand<number, "PhysicalAttackNum">

/** 魔法 攻撃 (数値) */
export type MagicAttackNum = Brand<number, "MagicAttackNum">

/** 物理 防御 (数値) */
export type PhysicalDefenseNum = Brand<number, "PhysicalDefenseNum">

/** 魔法 防御 (数値) */
export type MagicDefenseNum = Brand<number, "MagicDefenseNum">

/** 体力 (数値) */
export type HitPointNum = Brand<number, "HitPointNum">

/** % */
// type Rate = Brand<number, "Rate">

/** 特殊効果・ステータスに影響を与えるもの */
// type SpecialEffect_EffectStatuses = {
//   name: string
//   rate: Rate
//   rawText: string
// }
/** 特殊効果・ステータスに影響を与えない特殊効果 */
type SpecialEffect_Other = {
  rawText: string
}
/** 特殊効果 */
// type SpecialEffect = SpecialEffect_EffectStatuses | SpecialEffect_Other
export type SpecialEffect = SpecialEffect_Other

/**
 * @throws 装備が存在しなかった場合
 */
export const findEquipmentStrict = (equipmentId: EquipmentId): Equipment => {
  const found = data.find((e) => e.id === equipmentId)
  if (found == null) {
    throw new Error(`Logic failure. equipmentId: ${equipmentId} not found.`)
  }
  return found
}
