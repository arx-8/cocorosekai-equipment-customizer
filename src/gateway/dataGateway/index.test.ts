import { Equipment } from "src/domain/model/Equipment"
import { CastAny } from "src/types/tsUtils"

import { convertToItemsTable } from "."

const createTestData = (): readonly Equipment[] => {
  return [
    {
      id: 896902,
      imageUrl:
        "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/896900.png",
      equipCost: 64,
      name: "鬼灯鼻・じゅんろく",
      rawName: "鬼灯鼻・じゅんろく++",
      enhancementsNum: 2,
      attribute: "void",
      statuses: {
        hp: 0,
        magicAtk: 0,
        magicDef: 0,
        physicalAtk: 258,
        physicalDef: 0,
      },
      specialEffects: [
        {
          rawText: "クリティカル率が35%上昇する。",
        },
      ],
    },
    {
      id: 950000,
      imageUrl:
        "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/950000.png",
      equipCost: 10,
      name: "ヴィブロブレード",
      rawName: "ヴィブロブレード",
      enhancementsNum: 0,
      attribute: "void",
      statuses: {
        hp: 30,
        magicAtk: 0,
        magicDef: 0,
        physicalAtk: 48,
        physicalDef: 0,
      },
      specialEffects: [],
    },
    {
      id: 951602,
      imageUrl:
        "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/951600.png",
      equipCost: 51,
      name: "傲慢のアラーニェ",
      rawName: "傲慢のアラーニェ++",
      enhancementsNum: 2,
      attribute: "void",
      statuses: {
        hp: 0,
        magicAtk: 326,
        magicDef: 0,
        physicalAtk: 326,
        physicalDef: 0,
      },
      specialEffects: [],
    },
  ] as CastAny
}

describe("convertToItemsTable", () => {
  it("stock 0", () => {
    // ## Arrange ##
    const data = createTestData()
    const stockNums = {}
    const currentEquippedNums = {}
    const isCurrentRowCheckStock = false

    // ## Act ##
    const result = convertToItemsTable(
      data,
      stockNums,
      currentEquippedNums,
      isCurrentRowCheckStock
    )

    // ## Assert ##
    expect(result).toMatchSnapshot()
  })

  it("check ON, stock 0", () => {
    // ## Arrange ##
    const data = createTestData()
    const stockNums = {}
    const currentEquippedNums = {}
    const isCurrentRowCheckStock = true

    // ## Act ##
    const result = convertToItemsTable(
      data,
      stockNums,
      currentEquippedNums,
      isCurrentRowCheckStock
    )

    // ## Assert ##
    expect(result).toMatchSnapshot()
  })

  it("check ON, some stock", () => {
    // ## Arrange ##
    const data = createTestData()
    const stockNums = {
      896902: 1,
      951602: 22,
    }
    const currentEquippedNums = {
      896902: 1,
      951602: 21,
    }
    const isCurrentRowCheckStock = true

    // ## Act ##
    const result = convertToItemsTable(
      data,
      stockNums,
      currentEquippedNums,
      isCurrentRowCheckStock
    )

    // ## Assert ##
    expect(result).toMatchSnapshot()
  })
})
