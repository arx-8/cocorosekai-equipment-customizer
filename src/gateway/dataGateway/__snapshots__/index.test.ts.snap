// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToItemsTable check ON, some stock 1`] = `
Array [
  Object {
    "attribute": "void",
    "canEquip": false,
    "enhancementsNum": 2,
    "equipCost": 64,
    "id": 896902,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/896900.png",
    "name": "鬼灯鼻・じゅんろく",
    "rawName": "鬼灯鼻・じゅんろく++",
    "specialEffects": Array [
      Object {
        "rawText": "クリティカル率が35%上昇する。",
      },
    ],
    "specialEffectsText": "クリティカル率が35%上昇する。",
    "statuses": Object {
      "hp": 0,
      "magicAtk": 0,
      "magicDef": 0,
      "physicalAtk": 258,
      "physicalDef": 0,
    },
    "stockNum": 1,
  },
  Object {
    "attribute": "void",
    "canEquip": false,
    "enhancementsNum": 0,
    "equipCost": 10,
    "id": 950000,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/950000.png",
    "name": "ヴィブロブレード",
    "rawName": "ヴィブロブレード",
    "specialEffects": Array [],
    "specialEffectsText": "",
    "statuses": Object {
      "hp": 30,
      "magicAtk": 0,
      "magicDef": 0,
      "physicalAtk": 48,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
  Object {
    "attribute": "void",
    "canEquip": true,
    "enhancementsNum": 2,
    "equipCost": 51,
    "id": 951602,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/951600.png",
    "name": "傲慢のアラーニェ",
    "rawName": "傲慢のアラーニェ++",
    "specialEffects": Array [],
    "specialEffectsText": "",
    "statuses": Object {
      "hp": 0,
      "magicAtk": 326,
      "magicDef": 0,
      "physicalAtk": 326,
      "physicalDef": 0,
    },
    "stockNum": 22,
  },
]
`;

exports[`convertToItemsTable check ON, stock 0 1`] = `
Array [
  Object {
    "attribute": "void",
    "canEquip": false,
    "enhancementsNum": 2,
    "equipCost": 64,
    "id": 896902,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/896900.png",
    "name": "鬼灯鼻・じゅんろく",
    "rawName": "鬼灯鼻・じゅんろく++",
    "specialEffects": Array [
      Object {
        "rawText": "クリティカル率が35%上昇する。",
      },
    ],
    "specialEffectsText": "クリティカル率が35%上昇する。",
    "statuses": Object {
      "hp": 0,
      "magicAtk": 0,
      "magicDef": 0,
      "physicalAtk": 258,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
  Object {
    "attribute": "void",
    "canEquip": false,
    "enhancementsNum": 0,
    "equipCost": 10,
    "id": 950000,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/950000.png",
    "name": "ヴィブロブレード",
    "rawName": "ヴィブロブレード",
    "specialEffects": Array [],
    "specialEffectsText": "",
    "statuses": Object {
      "hp": 30,
      "magicAtk": 0,
      "magicDef": 0,
      "physicalAtk": 48,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
  Object {
    "attribute": "void",
    "canEquip": false,
    "enhancementsNum": 2,
    "equipCost": 51,
    "id": 951602,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/951600.png",
    "name": "傲慢のアラーニェ",
    "rawName": "傲慢のアラーニェ++",
    "specialEffects": Array [],
    "specialEffectsText": "",
    "statuses": Object {
      "hp": 0,
      "magicAtk": 326,
      "magicDef": 0,
      "physicalAtk": 326,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
]
`;

exports[`convertToItemsTable stock 0 1`] = `
Array [
  Object {
    "attribute": "void",
    "canEquip": true,
    "enhancementsNum": 2,
    "equipCost": 64,
    "id": 896902,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/896900.png",
    "name": "鬼灯鼻・じゅんろく",
    "rawName": "鬼灯鼻・じゅんろく++",
    "specialEffects": Array [
      Object {
        "rawText": "クリティカル率が35%上昇する。",
      },
    ],
    "specialEffectsText": "クリティカル率が35%上昇する。",
    "statuses": Object {
      "hp": 0,
      "magicAtk": 0,
      "magicDef": 0,
      "physicalAtk": 258,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
  Object {
    "attribute": "void",
    "canEquip": true,
    "enhancementsNum": 0,
    "equipCost": 10,
    "id": 950000,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/950000.png",
    "name": "ヴィブロブレード",
    "rawName": "ヴィブロブレード",
    "specialEffects": Array [],
    "specialEffectsText": "",
    "statuses": Object {
      "hp": 30,
      "magicAtk": 0,
      "magicDef": 0,
      "physicalAtk": 48,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
  Object {
    "attribute": "void",
    "canEquip": true,
    "enhancementsNum": 2,
    "equipCost": 51,
    "id": 951602,
    "imageUrl": "https://raw.githubusercontent.com/arx-8/cocorosekai-equipment-customizer/master/src/assets/equipments/951600.png",
    "name": "傲慢のアラーニェ",
    "rawName": "傲慢のアラーニェ++",
    "specialEffects": Array [],
    "specialEffectsText": "",
    "statuses": Object {
      "hp": 0,
      "magicAtk": 326,
      "magicDef": 0,
      "physicalAtk": 326,
      "physicalDef": 0,
    },
    "stockNum": 0,
  },
]
`;
